scipy/integrate/_ivp/lsoda.py

```diff
@@ -207,6 +227,38 @@ def _dense_output_impl(self):
 
         return LsodaDenseOutput(self.t_old, self.t, h, order, yh)
 
+    def _initialize_tcrit(self, t0, tcrit, t_bound):
+        if tcrit is None:
+            self.tcrit = np.asarray([t_bound])
+            return
+        tcrit = np.asarray(tcrit)
+        if self.direction == -1:
+            # integrate towards negative infinity, i.e. t0>t_bound
+            # remove critical times outside t0
+            tcrit = tcrit[tcrit<t0]
+            # remove critical times outside end
+            tcrit = tcrit[tcrit>t_bound]
+        elif self.direction == 1:
+            # integrate towards positive infinity, i.e. t0<t_bound
+            # remove critical times before start time
+            tcrit = tcrit[tcrit>t0]
+            # remove critical times after end time
+            tcrit = tcrit[tcrit<t_bound]
+        else:
+            # should be unreachable keep?
+            raise ValueError("Unexpected direction")
+        self.tcrit = np.append(tcrit, t_bound)
+
+    def _find_next_tcrit(self):
+        if self.tcrit.size == 0:
```

Comment:
From a performance perspective, these extra if statements will give extra overhead.  It would be better if the minimum length of `tcrit` is `1`, i.e. `t_bound` is always there.  Then we only need to handle this extra logic when a critical point is hit, which is infrequent.

Reply:
changed in https://github.com/scipy/scipy/pull/23385/commits/a96c26db331c01aba1a15f433865aac951053983